src/query/storages/fuse/src/table_functions/fuse_encodings/fuse_encoding.rs

```diff
@@ -17,151 +17,211 @@ use std::sync::Arc;
-            let segments = segments_io
-                .read_segments::<SegmentInfo>(chunk, false)
-                .await?;
-            for segment in segments {
-                let segment = segment?;
-                for block in segment.blocks.iter() {
-                    let column_meta = block.col_metas.get(&column_id).unwrap();
-                    let (offset, len) = column_meta.offset_length();
-                    let ranges = vec![(column_id, offset..(offset + len))];
-                    let read_settings = ReadSettings::from_ctx(&self.ctx)?;
-                    let merge_io_read_res = BlockReader::merge_io_read(
-                        &read_settings,
-                        self.table.operator.clone(),
-                        &block.location.0,
-                        ranges,
-                        true,
-                    )
+        let mut info = Vec::new();
+        for table in self.tables.clone() {
+            if matches!(table.storage_format, FuseStorageFormat::Parquet) {
+                continue;
+            }
+            let mut columns_info = vec![];
+            let snapshot = table.read_table_snapshot().await?;
+            if snapshot.is_none() {
+                continue;
+            }
+            let snapshot = snapshot.unwrap();
+
+            let segments_io =
+                SegmentsIO::create(self.ctx.clone(), table.operator.clone(), table.schema());
+
+            let chunk_size = self.ctx.get_settings().get_max_threads()? as usize * 4;
+
+            let schema = table.schema();
+            let fields = schema.fields();
+            for chunk in snapshot.segments.chunks(chunk_size) {
+                let segments = segments_io
+                    .read_segments::<SegmentInfo>(chunk, false)
                     .await?;
-                    let column_chunks = merge_io_read_res.columns_chunks()?;
-                    let pages = column_chunks
-                        .get(&column_id)
-                        .unwrap()
-                        .as_raw_data()
-                        .unwrap();
-                    let pages = std::io::Cursor::new(pages);
-                    let page_metas = column_meta.as_native().unwrap().pages.clone();
-                    let reader = NativeReader::new(pages, page_metas, vec![]);
-                    let this_pages_info = stat_simple(reader, arrow_field.clone())?.pages;
-                    pages_info.extend(this_pages_info);
+                for segment in segments {
+                    let segment = segment?;
+                    for block in segment.blocks.iter() {
+                        for field in fields {
+                            if field.is_nested() {
+                                continue;
+                            }
+                            let column_id = field.column_id;
+                            let arrow_field: Field = field.into();
+                            let column_meta = block.col_metas.get(&column_id).unwrap();
```

Comment:
unwrap is unsafe, it may be possible that column_id is not inside the col_metas.

Reply:
This was introduced by me, let me fix it later...